### PR TITLE
type1font.py fixes and test case

### LINKED ISSUE #4470: Issue encoding type1 fonts with python 3
Initially reported at https://github.com/ContinuumIO/anaconda-issues/issues/337

```python
import numpy as np
import matplotlib as mpl
import matplotlib.pylab as plt

mpl.rcdefaults()
# This doesn't work
 mpl.rcParams['font.serif'] = "Times, serif"
# This does
#mpl.rcParams['font.serif'] = "Bitstream Vera Serif, serif"
mpl.rcParams['text.usetex'] = "true"

x = np.random.normal(size=100)
y = np.random.normal(size=100)
plt.scatter(x, y)
plt.xlabel("$\gamma$")
plt.savefig("foo.png")
print("foo.png saved")
plt.savefig("bar.pdf")
print("bar.pdf saved")
```

traceback: 

```python

---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-10-3a778cad97aa> in <module>()
----> 1 exec(compile(open('/tmp/ipython-56520H0.py').read(), '/tmp/ipython-56520H0.py', 'exec')) # PYTHON-MODE

/tmp/ipython-56520H0.py in <module>()

/home/tcaswell/source/my_source/matplotlib/lib/matplotlib/pyplot.py in savefig(*args, **kwargs)
    577 def savefig(*args, **kwargs):
    578     fig = gcf()
--> 579     res = fig.savefig(*args, **kwargs)
    580     draw()   # need this if 'transparent=True' to reset colors
    581     return res

/home/tcaswell/source/my_source/matplotlib/lib/matplotlib/figure.py in savefig(self, *args, **kwargs)
   1507             self.set_frameon(frameon)
   1508 
-> 1509         self.canvas.print_figure(*args, **kwargs)
   1510 
   1511         if frameon:

/home/tcaswell/source/my_source/matplotlib/lib/matplotlib/backends/backend_qt5agg.py in print_figure(self, *args, **kwargs)
    158 
    159     def print_figure(self, *args, **kwargs):
--> 160         FigureCanvasAgg.print_figure(self, *args, **kwargs)
    161         self.draw()
    162 

/home/tcaswell/source/my_source/matplotlib/lib/matplotlib/backend_bases.py in print_figure(self, filename, dpi, facecolor, edgecolor, orientation, format, **kwargs)
   2223                 orientation=orientation,
   2224                 bbox_inches_restore=_bbox_inches_restore,
-> 2225                 **kwargs)
   2226         finally:
   2227             if bbox_inches and restore_bbox:

/home/tcaswell/source/my_source/matplotlib/lib/matplotlib/backends/backend_pdf.py in print_pdf(self, filename, **kwargs)
   2513                 file.endStream()
   2514             else:            # we opened the file above; now finish it off
-> 2515                 file.close()
   2516 
   2517 

/home/tcaswell/source/my_source/matplotlib/lib/matplotlib/backends/backend_pdf.py in close(self)
    552         self.endStream()
    553         # Write out the various deferred objects
--> 554         self.writeFonts()
    555         self.writeObject(self.alphaStateObject,
    556                          dict([(val[0], val[1])

/home/tcaswell/source/my_source/matplotlib/lib/matplotlib/backends/backend_pdf.py in writeFonts(self)
    648                 matplotlib.verbose.report('Writing Type-1 font', 'debug')
    649                 fonts[Fx] = self.embedTeXFont(filename,
--> 650                                               self.dviFontInfo[filename])
    651             else:
    652                 # a normal TrueType font

/home/tcaswell/source/my_source/matplotlib/lib/matplotlib/backends/backend_pdf.py in embedTeXFont(self, texname, fontinfo)
    713         t1font = type1font.Type1Font(fontinfo.fontfile)
    714         if fontinfo.effects:
--> 715             t1font = t1font.transform(fontinfo.effects)
    716         fontdict['BaseFont'] = Name(t1font.prop['FontName'])
    717 

/home/tcaswell/source/my_source/matplotlib/lib/matplotlib/type1font.py in transform(self, effects)
    334                 if six.PY3 and isinstance(value, int):
    335                     value = chr(value)
--> 336                 value = value.encode('latin-1')
    337                 buffer.write(value)
    338             result = buffer.getvalue()

AttributeError: 'bytes' object has no attribute 'encode'

```

### PR DESCRIPTION
**Pull request by jkseppan**
This started from reducing the Python 2 vs 3 special-casing in type1font as an alternative to #4472, then I added a test case using a Type-1 version of cmr10 as data and realized that the font transformations had been broken for quite a while.

Fixes #4470.

### CODE REVIEW

File: lib/matplotlib/type1font.py [hunk 1]
```diff
@@ -309,7 +309,7 @@ def suppress(tokens):
 
         while True:
             token, value = next(tokens)
-            if token == 'name' and value in table:
+            if token is cls._name and value in table:
```
Comment (tacaswell) [inline lib/matplotlib/type1font.py]: Why use `is` test instead of `==`?  That seems to be betting heavily on some cpython internals working as expected.

**Comment by tacaswell** [comment]
What is the license on cmr10?  If we are going to bundle a font, might as well include in in the main code so it is available to users?

**Comment by jkseppan** [comment]
> What is the license on cmr10? If we are going to bundle a font, might as well include in in the main code so it is available to users?

I chose cmr10 because it’s already included in lib/matplotlib/mpl-data/fonts/ttf, only in a different format. I think LICENSE/LICENSE_BAKOMA is supposed to cover this, but comments in the font file itself attribute the font to the AMS and mention the SIL license. We should check what exact fonts we do distribute and what their license terms are.

We don’t have full Type-1 font support, just this bare-bones parser to allow embedding TeX fonts in pdf files. The user’s TeX installation should provide all the Type-1 fonts needed, and this file is just test data.

**Comment by tacaswell** [comment]
sounds reasonable to me.

**Comment by tacaswell** [comment]
other than my concern about `is` rather than `==` this looks good to me.

File: lib/matplotlib/type1font.py (hunk 1 above)
Comment (jkseppan) [inline lib/matplotlib/type1font.py]: I introduced these sentinel objects to replace string comparisons, which were hard to get right across Python versions. The idea would be to use them like None. Perhaps this needs better comments, or one of the other enum patterns.

**Comment by jkseppan** [comment]
I started improving the tests and discovered an off-by-one error, so don't merge this quite yet.

File: lib/matplotlib/type1font.py (hunk 1 above)
Comment (tacaswell) [inline lib/matplotlib/type1font.py]: Ah, sorry :sheep: makes sense now.

Was this the stuff that was a long time broken do to `token` no longer being a string?

Comment (jkseppan) [inline lib/matplotlib/type1font.py]: Yes, the confusion between strings and bytes was exactly the problem.

**Comment by tacaswell** [comment]
It looks like this broke multi-page tiffs: https://travis-ci.org/matplotlib/matplotlib/jobs/66786896#L5834

File: lib/matplotlib/type1font.py [hunk 2]
```diff
@@ -136,17 +135,16 @@ def _split(self, data):
         # but if we read a pfa file, this part is already in hex, and
         # I am not quite sure if even the pfb format guarantees that
         # it will be in binary).
-        binary = b''.join([unichr(int(data[i:i + 2], 16)).encode('latin-1')
-                           for i in range(len1, idx, 2)])
+        binary = binascii.unhexlify(data[len1:idx+1])
```
Comment (tacaswell) [inline lib/matplotlib/type1font.py]: Something is going wrong in this line

**Pull request merged**